Make the 'history' and 'r' commands builtins

JohnoKing opened the pull request:
This pull request makes `history` and `r` commands builtins that aren't limited to interactive shells. All ksh commands can now be used without `unalias -a` removing them (`unalias -a` in `.kshrc` no longer removes `history` and `r`). With this change no more preset aliases exist, so the preset alias tables can be safely removed. The alias tree is now initialized in the same way as the function tree:
https://github.com/ksh93/ksh/blob/17f81ebedb440ed2297097446499a7016d5428e2/src/cmd/ksh93/sh/init.c#L1858
The `sh_inittree` code for comparison:
https://github.com/ksh93/ksh/blob/17f81ebedb440ed2297097446499a7016d5428e2/src/cmd/ksh93/sh/init.c#L1900

JohnoKing committed:
Make the 'history' and 'r' commands builtins

With this change no more preset aliases exist, so the preset alias
tables can be safely removed. All ksh commands can now be used without
'unalias -a' removing them, even in interactive shells. Additionally,
the history and r commands are no longer limited to being used in
interactive shells.

src/cmd/ksh93/bltins/hist.c:
- Implement the history and r commands as builtins. Also guarantee
  lflag is set to one by avoiding 'lflag++'.

src/cmd/ksh93/Makefile,
src/cmd/ksh93/Mamfile,
src/cmd/ksh93/sh/main.c,
src/cmd/ksh93/sh/init.c,
src/cmd/ksh93/data/aliases.c:
- Remove the table of predefined aliases because the last few have
  been removed. During init the alias tree is now initialized the
  same way as the function tree, rather than with a dummy shell table.

src/cmd/ksh93/bltins/typeset.c:
- Remove the now bugfix for unsetting predefined aliases because it
  is now a no-op. Aliases are no longer able to have the NV_NOFREE
  attribute.

src/cmd/ksh93/tests/alias.sh:
- Remove the regression test for unsetting predefined aliases since
  those no longer exist.

src/cmd/ksh93/sh.1:
- Remove the list of preset aliases since those no longer exist.
- Document history and r as builtins instead of preset aliases.

McDutchie commented on src/cmd/ksh93/bltins/hist.c:
```diff
@@ -58,6 +58,13 @@ int	b_hist(int argc,char *argv[], Shbltin_t *context)
 		errormsg(SH_DICT,ERROR_system(1),e_histopen);
+
+	/* 'history' and 'r' builtins */
+	if(argv[0][4] == 'o')		/* hist<o>ry */
```
This function may be called with an `argv[0]` that is shorter than 4 characters (`fc`, `r`), so this check is unsafe and causes unpredictable behaviour.

One way to fix it is:
```c
	if(argv[0][0] == 'h' && argv[0][4] == 'o')
```
as `argv[0][4]` is zero when called as `hist`.

Or we could just do:
```c
	if(!strcmp(argv[0],"history"))
```
After all, it's not as if we need to optimise performance here.

The check for `r` should be fine as is, as it checks the first character only and this function is never called by another command that starts with `r`.

JohnoKing replied:
I have applied the change that avoids `strcmp`.

JohnoKing committed:
Apply suggested change that avoids strcmp

JohnoKing committed:
Add comment noting argv[0][4] is zero when called as hist

McDutchie committed:
revert now-cosmetic changes introduced by 17f81ebe

This is just to avoid hard-to-read diffs when comparing across
multiple commits.

src/cmd/ksh93/sh/init.c:
- Rename sh_inittree() back to inittree().
- nv_init(): Move alias_tree init to same line as before.

McDutchie committed:
data/builtins.c: update sh_opthash[]
